File: wallendar-app/app/src/main/java/com/td/wallendar/di/DependenciesContainerLocator.java
package com.td.wallendar.di;

import android.content.Context;

import androidx.annotation.VisibleForTesting;

public class DependenciesContainerLocator {

    private static DependenciesContainer dependenciesContainer;

    private DependenciesContainerLocator() {
        // no-op
    }

    public static DependenciesContainer locateComponent(final Context context) {
        if (dependenciesContainer == null) {
            setComponent(new ProductionDependenciesContainer(context));
        }

        return dependenciesContainer;
    }

    @VisibleForTesting
    public static void setComponent(final DependenciesContainer dependenciesContainer) {
        DependenciesContainerLocator.dependenciesContainer = dependenciesContainer;
    }
}

File: wallendar-app/app/src/main/java/com/td/wallendar/di/DependenciesContainerLocator.kt
package com.td.wallendar.di

import android.content.Context
import androidx.annotation.VisibleForTesting

object DependenciesContainerLocator {
    private var dependenciesContainer: DependenciesContainer? = null
    fun locateComponent(context: Context): DependenciesContainer {
        if (dependenciesContainer == null) {
            setComponent(ProductionDependenciesContainer(context))
        }
        return dependenciesContainer!!
    }

    @VisibleForTesting
    fun setComponent(dependenciesContainer: DependenciesContainer) {
        DependenciesContainerLocator.dependenciesContainer = dependenciesContainer
    }
}